File: churn-demo/new/ledger.go
package ledger

import (
	"fmt"
	"math"
)

type Entry struct {
	Status string
	Amount float64
}

type Ledger struct {
	name          string
	previousName  string
	balance       float64
	limit         float64
	interestPaid  float64
	transfers     int
	pendingHolds  int
	clearedItems  int
	flaggedItems  int
	disputedItems int
	entries       []Entry
	summary       map[string]float64
}

func NewLedger(name string) *Ledger {
	return &Ledger{name: name, summary: map[string]float64{}}
}

func averageOf(sum float64, n int) float64 {
	if n == 0 {
		return 0
	}
	return sum / float64(n)
}

func (l *Ledger) Post(status string, amount float64) {
	l.entries = append(l.entries, Entry{Status: status, Amount: amount})
	l.balance = l.balance + amount
	l.log("posted")
}

func (l *Ledger) Credit(amount float64) {
	l.Post("posted", amount)
	l.clearedItems++
}

func (l *Ledger) Debit(amount float64) {
	l.Post("posted", -amount)
	l.clearedItems++
}

func (l *Ledger) Hold(amount float64) {
	l.Post("pending", amount)
	l.pendingHolds++
}

func (l *Ledger) Dispute(amount float64) {
	l.Post("disputed", amount)
	l.disputedItems++
	l.flaggedItems++
}

func (l *Ledger) Transfer(to *Ledger, amount float64) {
	l.Debit(amount)
	to.Credit(amount)
	l.transfers++
}

func (l *Ledger) ApplyInterest(rate float64) {
	interest := l.balance * rate
	l.Post("posted", interest)
	l.interestPaid += interest
}

func (l *Ledger) FlagLarge(threshold float64) int {
	n := 0
	for _, e := range l.entries {
		if math.Abs(e.Amount) > threshold {
			n++
		}
	}
	l.flaggedItems += n
	return n
}

func (l *Ledger) LargestEntry() float64 {
	best := 0.0
	for _, e := range l.entries {
		if math.Abs(e.Amount) > math.Abs(best) {
			best = e.Amount
		}
	}
	return best
}

func (l *Ledger) StatementLine(i int) string {
	e := l.entries[i]
	return fmt.Sprintf("%s:%.2f", e.Status, e.Amount)
}

func (l *Ledger) Rename(newName string) {
	l.previousName = l.name
	l.name = newName
	l.log("renamed")
}

func (l *Ledger) NetFlow() float64 {
	return l.sumSigned(+1) - l.sumSigned(-1)
}

func (l *Ledger) Reset() {
	l.entries = nil
	l.summary = map[string]float64{}
	l.balance = 0
	l.log("reset")
}

func (l *Ledger) sumSigned(sign int) float64 {
	sum := 0.0
	for _, e := range l.entries {
		if (e.Amount >= 0) == (sign > 0) {
			sum += math.Abs(e.Amount)
		}
	}
	return sum
}

func (l *Ledger) countStatus(status string) int {
	n := 0
	for _, e := range l.entries {
		if e.Status == status {
			n++
		}
	}
	return n
}

func (l *Ledger) SetLimit(limit float64) {
	l.limit = limit
	l.log("limit changed")
}

func (l *Ledger) OverLimit() bool {
	return l.balance < -l.limit
}

func (l *Ledger) Name() string {
	return l.name
}

func (l *Ledger) Balance() float64 {
	return l.balance
}

func (l *Ledger) EntryCount() int {
	return len(l.entries)
}

func (l *Ledger) SummaryValue(key string) float64 {
	return l.summary[key]
}

func (l *Ledger) Reconcile() {
	opening := l.balance
	credits := l.sumSigned(+1)
	debits := l.sumSigned(-1)
	closing := opening + credits - debits
	drift := closing - l.balance
	posted := l.countStatus("posted")
	pending := l.countStatus("pending")
	disputed := l.countStatus("disputed")
	total := posted + pending + disputed
	average := averageOf(credits+debits, total)
	l.summary["opening"] = opening
	l.summary["credits"] = credits
	l.summary["debits"] = debits
	l.summary["closing"] = closing
	l.summary["drift"] = drift
	l.summary["posted"] = float64(posted)
	l.summary["pending"] = float64(pending)
	l.summary["disputed"] = float64(disputed)
	l.summary["average"] = average
	l.balance = closing
}

func (l *Ledger) log(m string) {}

File: churn-demo/new/ledger.ts
interface Entry { status: string; amount: number; }

class Ledger {
    name: string;
    previousName: string;
    balance: number;
    limit: number;
    interestPaid: number;
    transfers: number;
    pendingHolds: number;
    clearedItems: number;
    flaggedItems: number;
    disputedItems: number;
    entries: Entry[];
    summary: Map<string, number>;
    lastLog: string;
    constructor(name: string) {
        this.name = name;
        this.previousName = "";
        this.balance = 0;
        this.limit = 0;
        this.interestPaid = 0;
        this.transfers = 0;
        this.pendingHolds = 0;
        this.clearedItems = 0;
        this.flaggedItems = 0;
        this.disputedItems = 0;
        this.entries = [];
        this.summary = new Map();
        this.lastLog = "";
    }

    log(m: string): void {
        this.lastLog = m;
    }

    post(status: string, amount: number): void {
        this.entries.push({ status, amount });
        this.balance = this.balance + amount;
        this.log("posted");
    }

    credit(amount: number): void {
        this.post("posted", amount);
        this.clearedItems += 1;
    }

    debit(amount: number): void {
        this.post("posted", -amount);
        this.clearedItems += 1;
    }

    hold(amount: number): void {
        this.post("pending", amount);
        this.pendingHolds += 1;
    }

    dispute(amount: number): void {
        this.post("disputed", amount);
        this.disputedItems += 1;
        this.flaggedItems += 1;
    }

    transfer(to: Ledger, amount: number): void {
        this.debit(amount);
        to.credit(amount);
        this.transfers += 1;
    }

    applyInterest(rate: number): void {
        const interest = this.balance * rate;
        this.post("posted", interest);
        this.interestPaid += interest;
    }

    flagLarge(threshold: number): number {
        let n = 0;
        for (const e of this.entries) {
            if (Math.abs(e.amount) > threshold) n += 1;
        }
        this.flaggedItems += n;
        return n;
    }

    largestEntry(): number {
        let best = 0;
        for (const e of this.entries) {
            if (Math.abs(e.amount) > Math.abs(best)) best = e.amount;
        }
        return best;
    }

    statementLine(i: number): string {
        const e = this.entries[i];
        return e.status + ":" + String(e.amount);
    }

    rename(newName: string): void {
        this.previousName = this.name;
        this.name = newName;
        this.log("renamed");
    }

    netFlow(): number {
        return this.sumSigned(+1) - this.sumSigned(-1);
    }

    reset(): void {
        this.entries = [];
        this.summary = new Map();
        this.balance = 0;
        this.log("reset");
    }

    sumSigned(sign: number): number {
        let sum = 0;
        for (const e of this.entries) {
            if ((e.amount >= 0) === (sign > 0)) sum += Math.abs(e.amount);
        }
        return sum;
    }

    countStatus(status: string): number {
        let n = 0;
        for (const e of this.entries) {
            if (e.status === status) n += 1;
        }
        return n;
    }

    setLimit(limit: number): void {
        this.limit = limit;
        this.log("limit changed");
    }

    overLimit(): boolean {
        return this.balance < -this.limit;
    }

    entryCount(): number {
        return this.entries.length;
    }

    summaryValue(key: string): number {
        return this.summary.get(key) ?? 0;
    }

    reconcile(): void {
        const opening = this.balance;
        const credits = this.sumSigned(+1);
        const debits = this.sumSigned(-1);
        const closing = opening + credits - debits;
        const drift = closing - this.balance;
        const posted = this.countStatus("posted");
        const pending = this.countStatus("pending");
        const disputed = this.countStatus("disputed");
        const total = posted + pending + disputed;
        const average = total > 0 ? (credits + debits) / total : 0;
        this.summary.set("opening", opening);
        this.summary.set("credits", credits);
        this.summary.set("debits", debits);
        this.summary.set("closing", closing);
        this.summary.set("drift", drift);
        this.summary.set("posted", posted);
        this.summary.set("pending", pending);
        this.summary.set("disputed", disputed);
        this.summary.set("average", average);
        this.balance = closing + this.interestPaid;
    }

    resetCounters(): void { this.pendingHolds = 0; this.clearedItems = 0; this.flaggedItems = 0; this.disputedItems = 0; }
}

export { Ledger };
